tools/hidl/AccessMapGenerator.hpp

```diff
@@ -0,0 +1,1003 @@
+        {
+        }
+        virtual void reportInfo(const char* fmt, va_list& args) const
+        {
+        }
+        virtual void reportDebug(const char* fmt, va_list& args) const
+        {
+        }
+    };
+
+protected:
+    struct StringICompare
+    {
+        bool operator () (const std::string& lhs, const std::string& rhs) const
+        {
+            return (stricmp(lhs.c_str(), rhs.c_str()) < 0);
+        }
+    };
+
+    using Scopes = std::map<int, std::string>;
+    using Variables = std::map<std::string, std::string, StringICompare>;
+    using Tokens = std::list<std::string>;
+
+    struct Term
+    {
+        int sourceScope = ScopeMapper::Invalid;  // defining scope
+        int affectsScope = ScopeMapper::Omitted; // scope affected by the term; used only for exclusions
+        std::string feature;                     // feature name
+        TLevel level = TLevel();                 // access level; ignored by exclusions
+        bool exclusion = false;                  // true --> this invalidates lower priority entries
+        bool suppression = false;                // true --> this invalidates lower priority entries and affirms the absence of security
+        bool deferral = false;                   // true --> this suggests that security should be applied within method handlers
+        bool all = false;                        // true --> suppression or deferral applies to all
+
+        bool featureMatches(const char* name) const
+        {
+            if (nullptr == name)
+                return false;
+            bool result = false;
+            if (exclusion)
+            {
+                if (feature.empty() || streq(feature, "*"))
+                    result = true;
+                else if (strieq(feature, name))
+                    result = true;
+                // What if name is "" or "*"?
+            }
+            else
+            {
+                result = strieq(feature, name);
+            }
+            return result;
+        }
+        inline bool featureMatches(const std::string& name) const
+        {
+            return featureMatches(name.c_str());
+        }
+    };
+    using Terms = std::list<Term>;
+    using TermPtrs = std::list<Term*>;
+
+public:
+    TAccessMapGenerator(const ScopeMapper& scopeMapper, const LevelMapper& levelMapper, const Reporter& reporter)
+        : m_scopeMapper(scopeMapper)
+        , m_levelMapper(levelMapper)
+        , m_reporter(reporter)
+    {
+    }
+
+    bool setRequireSecurityAffirmation(bool require)
+    {
+        bool wasRequired = m_requireSecurityAffirmation;
+        m_requireSecurityAffirmation = require;
+        return wasRequired;
+    }
+
+    bool setDefaultSecurity(const char* token)
+    {
+        bool changed = false;
+
+        if (isEmptyString(token))
+        {
+            if (m_useDefaultSecurity)
+            {
+                m_useDefaultSecurity = true;
```

Comment:
If it is already true why set it and also why set changed=true? If the value was already true it hasn't been changed.

Reply:
It's a mistake. An empty string is meant to disable the use of default security, so the assignment should be to false.